keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,215 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Kubernetes imposes some restrictions on the size of the Service Cluster CIDR,
+for IPv4 the prefix size is limited to /12, however, for IPv6 it is limited to
+/112 or fewer.
+This restriction is causing issues for IPv6 users, since /64 is the standard
+and minimum recommended prefix length.
+
+## Motivation
+
+IPv6 users should be able to configure Kubernetes cluster following the best
+networking practices, to avoid any issues caused by the use of bad assignment
+policies.
+
+https://tools.ietf.org/html/rfc7421
+https://www.ripe.net/publications/docs/ripe-738
+
+### Goals
+
+Allow to use a /64 prefix for the Service CIDR Subnet
+
+### Non-Goals
+
+Any architectural or drastic change
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes admin that wants to use IPv6, I want to be able to follow the IPv6 address
+planning best practices, being able to assign /64 prefixes to my end subnets.
+
+### Notes/Constraints/Caveats
+
+There is no much adoption for Kubernetes IPv6 clusters right now, only a few users claim to be running
+IPv6 clusters in production, but with the upcoming of Dual Stack we can observe that is getting more traction.
+A common complain from new users with an important networking background is the limitation to not being
+able to use a /64 for the service subnet.
+https://github.com/kubernetes/kubernetes/issues/92451
+
+Right now this is a cosmetic change and it will not have too much impact, it will be more
+difficult to do it once there is more IPv6 adoption.
+
+## Design Details
+
+The service IP allocator is implemented using a bitmap to store and track the assigned ClusterIPs.
+This bitmap is stored in `etcd` to provide consistency, so multiple `kube-apiserver` can use it as 
+single source of truth, and assign IP addresses without conflicts.
+To avoid performance issues because of the size of the bitmap, the `kube-apiserver` limits
+maximum subnet size for IPv4 and IPv6.
+However, there is a inconsistency here, because the IPv4 subnet size is limited to /12, 
+but not the bitmap size, so we can have a bitmap of 2^20 bits in IPv4.
+In the other hand, for IPv6 we always limit the bitmap size to 2^16 bits, so despite allowing a 
+/112 prefix we only allocate from the low 16 bits of the subnet.
+
+```
+<<[UNRESOLVED max bitmap size ]>>
+
+1. keep current status for bitmap size max: IPv4 = 2^20 , IPv6 = 2^16
+2. consolidate the max size for both IP families
+  2.1 2^16
+  2.2 2^20
+  2.3 ...
+
+<<[/UNRESOLVED]>>
+```
+
+The solution proposed consist in:
+
+Relax only the `kube-apiserver` validation without any further changes in the ipallocator logic,
+so the user can provide a /64 prefix for IPv6 but only allocate out of the lowest 2^16 bits.
```

Comment:
Back-of-envelope math.  If the user specifies a large CIDR, we could allow a larger subset (e.g. 2^22 bits, 512KiB), but that's pretty insignificant compared to 2^64.

If we thought that reserving specific IPs was important, we would change the structure to allow arbitrary specific values plus a bitmap of allocated values.  E.g. 2^16 bits (8KiB) of allocation + list of up to 10K offsets from the CIDR (8 bytes each, 80 KiB).  But as you say, that would require retooling the structure that is stored and careful rollout.

If we did that, we could also move to a model that stored multiple chunks of data and even allowed multiple CIDRs.  It's a lot of work and not very high demand.  So I guess I am OK with this proposal, and if we decide later to allow discrete allocations, we can work on converting the structures.

Reply:
When you say lowest bits, I assume you mean leading 0s?  Maybe an example.

Reply:
> When you say lowest bits, I assume you mean leading 0s? Maybe an example.

heh didn't know asciiflow worked for markdown,
what about something as this?


      +---------------------------------------+
      | 2001:db8:1234:5678:0000:0000:0000:0000|
      +-------------------+--------------+----+
                          |              |
                          |--->          |
                         /64             |
                        subnet           |
                                         +---->
                                       /112
                                     ipallocator
                                        16 bits

- subnet range 
`2001:db8:1234:5678:0000:0000:0000:0000 - 2001:0db8:1234:5678:ffff:ffff:ffff:ffff`
- kubernetes allocator subnet range 
`2001:db8:1234:5678:0000:0000:0000:0000 - 2001:0db8:1234:5678:0000:0000:0000:ffff`

> if we decide later to allow discrete allocations, we can work on converting the structures.

yeah, nowadays /64 is a fair request based on real facts and a few users are demanding it
OTHO discrete allocation in the whole range is a guess or a nice to have at this moment ...